File: KSC2015_ProblemSet_n_Solution/Competition2015/04_KISTI/Code/C/mpi/poisson_mpi.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <mpi.h>
#include "matrixconstructor.h"
#include "cgsolver_mpi.h"

#define PI 3.141592653589793

int main(int argc, char **argv) {

	int numgrid_x, numgrid_y, firstgrid_x, lastgrid_x;
	int firstrow, lastrow, matrixDOF, globalmatrixDOF;  
	int count, ii, jj, kk, maxiteration, myrank, ncpus;  
	double length_x, length_y, gridsize, tolerance;
	double *poissonmatrix, *rhs, *solution, *coordinate;
	double elapsed_time;
	char myfilename[256];
	FILE *fp;

        MPI_Init(&argc, &argv);
        MPI_Comm_rank(MPI_COMM_WORLD, &myrank);
        MPI_Comm_size(MPI_COMM_WORLD, &ncpus);

	if(myrank == 0)
		printf("[Poisson] MPI initialized.\n");

	elapsed_time = -MPI_Wtime();

	maxiteration = 20000;
	tolerance = 1.0e-10;
	gridsize = 0.01;
	length_x = 1.0;
	length_y = 1.0;

	numgrid_x = (int)round(length_x/gridsize) + 1;
	numgrid_y = (int)round(length_y/gridsize) + 1;
	globalmatrixDOF = numgrid_x*numgrid_y;

        firstgrid_x = (numgrid_x*myrank/ncpus) + 1;
	lastgrid_x = numgrid_x*(myrank+1)/ncpus;
        firstrow = (firstgrid_x-1)*numgrid_y + 1;
        lastrow = lastgrid_x*numgrid_y;
	matrixDOF = lastrow - firstrow + 1;

	if(myrank == 0)
		printf("[Poisson] Geometry and matrix size initialized.\n");

	coordinate = (double*)malloc(matrixDOF*2*sizeof(double));
	rhs = (double*)malloc(matrixDOF*sizeof(double));

	count = 0;

	for(ii=firstgrid_x; ii<=lastgrid_x; ii++)
	{
		for(jj=1; jj<=numgrid_y; jj++)
		{
			coordinate[2*count] = (ii-1)*gridsize;
			coordinate[2*count+1] = (jj-1)*gridsize;
			rhs[count] = sin(coordinate[2*count]/length_x*PI)*sin(coordinate[2*count+1]/length_y*PI)*gridsize*gridsize;
			count++;
		} 
	}

	if(myrank == 0)	
		printf("[Poisson] Geometry and rhs constructed.\n");

	poissonmatrix = (double*) malloc(matrixDOF*globalmatrixDOF*sizeof(double)); 

	for(ii=0; ii<matrixDOF*globalmatrixDOF; ii++)	
		poissonmatrix[ii] = 0.0;

	construct_poissonmatrix(firstrow, firstgrid_x, lastgrid_x, numgrid_x, numgrid_y, poissonmatrix);

	if(myrank == 0)
		printf("[Poisson] Poisson matrix constructed.\n");

	solution = (double*) malloc(matrixDOF*sizeof(double));

	for(ii=0; ii<matrixDOF; ii++)
		solution[ii] = 1.0;

	if(myrank == 0)
		printf("[Poisson] Start solving equations.\n");

	cgsolver(matrixDOF, myrank, ncpus, poissonmatrix, rhs, solution, maxiteration, tolerance);

	if(myrank == 0)
	{
		printf("[Poisson] Solution obtained.\n");
		system("rm -rf result\n");
		system("mkdir result\n");
	}

	MPI_Barrier(MPI_COMM_WORLD);

	sprintf(myfilename,"./result/solution.dat");
	
	for(ii=0; ii<=ncpus; ii++)
	{
		if(myrank == ii)
		{
			fp = fopen(myfilename, "a");
			count = 0;
	
			for(jj=firstgrid_x; jj<=lastgrid_x; jj++)
			{
				for(kk=1; kk<=numgrid_y; kk++)
				{
					fprintf(fp,"%9.5f%10.5f%13.6f\n", coordinate[2*count], coordinate[2*count+1], solution[count]);
					count++;
				}
			}

			fclose(fp);
		}

		MPI_Barrier(MPI_COMM_WORLD);
	}

	if(myrank == 0)
		printf("[Poisson] Solution printed.\n");

	free(solution);
	free(poissonmatrix);
	free(rhs);
	free(coordinate);

	if(myrank == 0)
		printf("[Poisson] Memory deallocated.\n");

	elapsed_time += MPI_Wtime();	

	if(myrank == 0)
		 printf("[Poisson] Finalizing MPI - Computing finished in %e (secs).\n", elapsed_time);

	MPI_Finalize();
	return 0;
}

File: KSC2015_ProblemSet_n_Solution/Competition2015/04_KISTI/Code/Fortran/mpi/poisson_mpi.f90
program main

      use mpi

      implicit none

      real(8), parameter :: PI = 3.1415926535897932384626
      integer(4) :: numgrid_x, numgrid_y, firstgrid_x, lastgrid_x
      integer(4) :: firstrow, lastrow, matrixDOF, globalmatrixDOF
      integer(4) :: count, ii, jj, kk, maxiteration, ierr, myrank, ncpus
      real(8) :: length_x, length_y, gridsize, tolerance
      real(8), allocatable :: poissonmatrix(:), rhs(:), solution(:), coordinate(:)
      real(8) :: elapsed_time
      character(256) ::  myfilename

      call MPI_Init(ierr)
      call MPI_Comm_size(MPI_COMM_WORLD,ncpus,ierr)
      call MPI_Comm_rank(MPI_COMM_WORLD,myrank,ierr)

      if(myrank.EQ.0) then
          write(*,*) '[Poisson] MPI initialized.'
      endif

      elapsed_time = -MPI_Wtime()

      maxiteration = 20000
      tolerance = 1.0e-10
      gridsize = 0.01
      length_x = 1.0
      length_y = 1.0

      numgrid_x = nint(length_x/gridsize) + 1
      numgrid_y = nint(length_y/gridsize) + 1
      globalmatrixDOF = numgrid_x * numgrid_y

      firstgrid_x = (numgrid_x * myrank / ncpus) + 1
      lastgrid_x = numgrid_x * (myrank + 1) / ncpus
      firstrow = (firstgrid_x - 1) * numgrid_y + 1
      lastrow = lastgrid_x * numgrid_y
      matrixDOF =  lastrow - firstrow + 1

      if(myrank.EQ.0) then
          write(*,*) '[Poisson] Geometry and matrix size initialized.'
      endif

      allocate( coordinate(matrixDOF*2) ) 
      allocate( rhs(matrixDOF) )

      count = 1

      do ii = firstgrid_x, lastgrid_x
          do jj = 1, numgrid_y
              coordinate(2*count-1) = (ii-1)*gridsize;
              coordinate(2*count) = (jj-1)*gridsize;
              rhs(count) = sin(coordinate(2*count-1)/length_x*PI)*sin(coordinate(2*count)/length_y*PI)*gridsize*gridsize
              count = count+1
          enddo
      enddo

      if(myrank.EQ.0) then
          write(*,*) '[Poisson] Geometry and rhs constructed.'
      endif

      allocate( poissonmatrix(matrixDOF * globalmatrixDOF) )

      do ii=1, matrixDOF * globalmatrixDOF
          poissonmatrix(ii) = 0.0
      enddo

      call construct_poissonmatrix(firstrow, firstgrid_x, lastgrid_x, numgrid_x, numgrid_y, poissonmatrix)

      if(myrank.EQ.0) then
          write(*,*) '[Poisson] Poisson matrix constructed.'
      endif

      allocate( solution(matrixDOF) )

      do ii=1, matrixDOF
          solution(ii) = 1.0;
      enddo

      if(myrank.EQ.0) then
          write(*,*) '[Poisson] Start solving equations.'
      endif

      call cgsolver(matrixDOF, myrank, ncpus, poissonmatrix, rhs, solution, maxiteration, tolerance)

      if(myrank.EQ.0) then
          write(*,*) '[Poisson] Solution obtained.'
          call system('rm -rf result')
          call system('mkdir result')
      endif

      call MPI_Barrier(MPI_COMM_WORLD,ierr)
      
      do ii = 0, ncpus-1
          if(myrank.EQ.ii) then
              open(11, file='./result/solution.dat', form='formatted',position='append')
              count = 1

              do jj=firstgrid_x, lastgrid_x
                  do kk=1, numgrid_y
                      write(11,101) coordinate(2*count-1), coordinate(2*count), solution(count)
                      count = count + 1
                  enddo
              enddo
              close(11)
          endif
          call MPI_Barrier(MPI_COMM_WORLD,ierr)
      enddo
101   format(2(f9.5,x),f12.6)

      if(myrank.EQ.0) then
          write(*,*) '[Poisson] Solution printed.'
      endif

      deallocate(solution)
      deallocate(poissonmatrix)
      deallocate(rhs)
      deallocate(coordinate)

      if(myrank.EQ.0) then
          write(*,*) '[Poisson] Memory deallocated.'
      endif

      elapsed_time = elapsed_time + MPI_Wtime()

102   format('[Poisson] Finalizing MPI - Computing finished in ',e15.9,' (secs).')
      if(myrank.EQ.0) then
          write(*,102) elapsed_time
      endif

      call MPI_Finalize(ierr)

end program

